scimma_admin/hopskotch_auth/models.py

```diff
@@ -202,3 +207,312 @@ def rand_username(owner: User) -> str:
+                repairOrDeletePermission(cred_perm, 
+                                         lambda other_group_perm: other_group_perm==permission,
+                                         user_memberships, user_group_perms)
+
+    return True
+    
+
+class CredentialKafkaPermission(models.Model):
+    principal = models.ForeignKey(
+        SCRAMCredentials,
+        on_delete=models.CASCADE,
+    )
+    # individual crentials derive their permissions from group permissions, 
+    # so we track that relationship in order to coordinate changes
+    parent = models.ForeignKey(
+        GroupKafkaPermission,
+        on_delete=models.CASCADE
+    )
+    topic = models.ForeignKey(
+        KafkaTopic,
+        on_delete=models.CASCADE,
+    )
+    operation = enum.EnumField(
+        KafkaOperation,
+        editable=False,
+    )
+
+
+# This needs to be a character which is valid neither in Kafka topic names nor in enum names
+cred_perm_encoding_separator=':'
+
+
+def encode_cred_permission(parent_id, topic_id, operation) -> str:
+    return str(parent_id)+cred_perm_encoding_separator+str(topic_id)+cred_perm_encoding_separator+str(operation)
+
+
+# returns tuples of (parent ID, topic ID, operation type)
+def decode_cred_permission(encoded: str):
+    pattern = re.compile("([^"+cred_perm_encoding_separator+"]+)"+cred_perm_encoding_separator+"([^"+cred_perm_encoding_separator+"]+)"+cred_perm_encoding_separator+"([^"+cred_perm_encoding_separator+"]+)")
+    match = re.fullmatch(pattern, encoded)
+    if match is None or len(match.groups())!=3:
+        raise ValueError("Invalid encoded credential permission")
+    try:
+        op = KafkaOperation[match[3]]
+    except KeyError:
+        raise ValueError("Invalid encoded credential permission")
+    return (match[1],match[2],op)
+
+
+# determine whether credential permission can be validly derived from the given group permission
+def supportingPermission(group_perm,cred_perm):
+    if cred_perm.topic != group_perm.topic:
+        return False
+    return group_perm.operation == KafkaOperation.All \
+      or cred_perm.operation == group_perm.operation
+
+
+# returns tuples of (parent ID, topic ID, topic name, perm type)
+def all_permissions_for_user(user):
+    possible_permissions = []
+    for membership in user.groupmembership_set.all():
+        group = membership.group
+        group_permissions = GroupKafkaPermission.objects.filter(principal=group).select_related('topic')
+        for permission in group_permissions:
+            if permission.operation==KafkaOperation.All:
+                for subpermission in KafkaOperation.__members__:
+                    possible_permissions.append((permission.id,permission.topic.id,permission.topic.name,subpermission))
+            else:
+                possible_permissions.append((permission.id,permission.topic.id,permission.topic.name,permission.operation))
+    # sort and eliminate duplicates
```

Comment:
Why would there be duplicates? It's fine if its just to deal with bugs, but it looked to me above like we forbid duplicates when creating permissions.

Reply:
There can be duplicates, for the purposes of this function, because the user may belong to multiple groups which eaach have access to a given topic. The group permission records are not identical, but the options they would provide to the user are, so we want to simplify.

Reply:
Good point!